File: bigclaw-go/scripts/e2e/mixed_workload_matrix.py
#!/usr/bin/env python3
import argparse
import json
import os
import pathlib
import shutil
import socket
import subprocess
import sys
import tempfile
import time
import urllib.request


def http_json(url, method='GET', payload=None, timeout=30):
    data = None if payload is None else json.dumps(payload).encode('utf-8')
    req = urllib.request.Request(url, data=data, method=method, headers={'Content-Type': 'application/json'})
    with urllib.request.urlopen(req, timeout=timeout) as resp:
        return json.loads(resp.read().decode('utf-8'))


def wait_for_health(base_url, attempts=60, interval=1.0):
    last_error = None
    for _ in range(attempts):
        try:
            if http_json(base_url + '/healthz').get('ok'):
                return
        except Exception as exc:
            last_error = exc
        time.sleep(interval)
    raise RuntimeError(f'service did not become healthy: {last_error}')


def reserve_local_base_url():
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
        sock.bind(('127.0.0.1', 0))
        sock.listen(1)
        port = sock.getsockname()[1]
    return f'http://127.0.0.1:{port}', f'127.0.0.1:{port}'


def build_autostart_env():
    env = os.environ.copy()
    state_dir = pathlib.Path(tempfile.mkdtemp(prefix='bigclawd-mixed-state-'))
    base_url, http_addr = reserve_local_base_url()
    env['BIGCLAW_HTTP_ADDR'] = http_addr
    env['BIGCLAW_QUEUE_BACKEND'] = env.get('BIGCLAW_QUEUE_BACKEND', 'sqlite') or 'sqlite'
    env['BIGCLAW_QUEUE_SQLITE_PATH'] = str(state_dir / 'queue.db')
    env['BIGCLAW_AUDIT_LOG_PATH'] = str(state_dir / 'audit.jsonl')
    env['BIGCLAW_SERVICE_NAME'] = env.get('BIGCLAW_SERVICE_NAME', 'bigclawd-mixed')
    return env, base_url, state_dir


def start_bigclawd(go_root, env):
    log_path = pathlib.Path(tempfile.mkstemp(prefix='bigclawd-mixed-', suffix='.log')[1])
    log_file = log_path.open('w')
    process = subprocess.Popen(['go', 'run', './cmd/bigclawd'], cwd=go_root, stdout=log_file, stderr=subprocess.STDOUT, env=env)
    return process, log_path


def terminal(state):
    return state in {'succeeded', 'dead_letter', 'cancelled', 'failed'}


def wait_task(base_url, task_id, timeout_seconds):
    deadline = time.time() + timeout_seconds
    while time.time() < deadline:
        status = http_json(base_url + f'/tasks/{task_id}')
        if terminal(status['state']):
            return status
        time.sleep(0.5)
    raise TimeoutError(task_id)


def fetch_events(base_url, task_id):
    return http_json(base_url + f'/events?task_id={task_id}&limit=100')['events']


def routed_event(events):
    for event in events:
        if event.get('type') == 'scheduler.routed':
            return event
    return None


def default_tasks(timestamp):
    return [
        {
            'name': 'local-default',
            'expected_executor': 'local',
            'expectation_label': 'default low/medium risk route',
            'expectation_rule': 'unscoped low-risk tasks stay on the local executor',
            'task': {
                'id': f'mixed-local-{timestamp}',
                'trace_id': f'mixed-local-{timestamp}',
                'title': 'Mixed workload local default',
                'entrypoint': 'echo local default',
                'metadata': {'scenario': 'mixed-workload', 'profile': 'local-default'},
            },
        },
        {
            'name': 'browser-auto',
            'expected_executor': 'kubernetes',
            'expectation_label': 'browser tool route',
            'expectation_rule': 'browser-required workloads should route to kubernetes',
            'task': {
                'id': f'mixed-browser-{timestamp}',
                'trace_id': f'mixed-browser-{timestamp}',
                'title': 'Mixed workload browser auto-route',
                'required_tools': ['browser'],
                'container_image': 'alpine:3.20',
                'entrypoint': 'echo browser via kubernetes',
                'metadata': {'scenario': 'mixed-workload', 'profile': 'browser-auto'},
            },
        },
        {
            'name': 'gpu-auto',
            'expected_executor': 'ray',
            'expectation_label': 'gpu tool route',
            'expectation_rule': 'gpu-required workloads should route to ray',
            'task': {
                'id': f'mixed-gpu-{timestamp}',
                'trace_id': f'mixed-gpu-{timestamp}',
                'title': 'Mixed workload gpu auto-route',
                'required_tools': ['gpu'],
                'entrypoint': "python -c \"print('gpu via ray')\"",
                'metadata': {'scenario': 'mixed-workload', 'profile': 'gpu-auto'},
            },
        },
        {
            'name': 'high-risk-auto',
            'expected_executor': 'kubernetes',
            'expectation_label': 'high-risk isolation route',
            'expectation_rule': 'high-risk workloads should route to the isolated kubernetes executor',
            'task': {
                'id': f'mixed-risk-{timestamp}',
                'trace_id': f'mixed-risk-{timestamp}',
                'title': 'Mixed workload high-risk auto-route',
                'risk_level': 'high',
                'container_image': 'alpine:3.20',
                'entrypoint': 'echo high risk via kubernetes',
                'metadata': {'scenario': 'mixed-workload', 'profile': 'high-risk-auto'},
            },
        },
        {
            'name': 'required-ray',
            'expected_executor': 'ray',
            'expectation_label': 'explicit required executor route',
            'expectation_rule': 'required_executor=ray must override automatic routing',
            'task': {
                'id': f'mixed-required-ray-{timestamp}',
                'trace_id': f'mixed-required-ray-{timestamp}',
                'title': 'Mixed workload explicit ray',
                'required_executor': 'ray',
                'entrypoint': "python -c \"print('required ray')\"",
                'metadata': {'scenario': 'mixed-workload', 'profile': 'required-ray'},
            },
        },
    ]


def write_report(go_root, report_path, payload):
    output_path = pathlib.Path(go_root) / report_path
    output_path.parent.mkdir(parents=True, exist_ok=True)
    output_path.write_text(json.dumps(payload, ensure_ascii=False, indent=2) + '\n', encoding='utf-8')


def copy_artifact(source, destination):
    source_path = pathlib.Path(source)
    if not source_path.exists():
        return ''
    destination.parent.mkdir(parents=True, exist_ok=True)
    shutil.copy2(source_path, destination)
    return str(destination)


def relpath(path, go_root):
    try:
        return str(pathlib.Path(path).resolve().relative_to(pathlib.Path(go_root).resolve()))
    except ValueError:
        return str(path)


def event_summary(event):
    payload = event.get('payload') if isinstance(event.get('payload'), dict) else {}
    return {
        'id': event.get('id', ''),
        'type': event.get('type', ''),
        'timestamp': event.get('timestamp', ''),
        'executor': payload.get('executor', ''),
        'reason': payload.get('reason', ''),
        'message': payload.get('message', ''),
    }


def build_task_result(spec, status, events):
    routed = routed_event(events)
    routed_executor = None
    routed_reason = None
    if routed and isinstance(routed.get('payload'), dict):
        routed_executor = routed['payload'].get('executor')
        routed_reason = routed['payload'].get('reason')
    latest_event_type = status['latest_event']['type'] if status.get('latest_event') else ''
    return {
        'name': spec['name'],
        'task_id': spec['task']['id'],
        'trace_id': spec['task']['trace_id'],
        'expectation': {
            'executor': spec['expected_executor'],
            'label': spec['expectation_label'],
            'rule': spec['expectation_rule'],
        },
        'observed': {
            'routed_executor': routed_executor,
            'routed_reason': routed_reason,
            'final_state': status['state'],
            'latest_event_type': latest_event_type,
        },
        'event_highlights': [event_summary(event) for event in events],
        'event_types': [event.get('type', '') for event in events],
        'event_count': len(events),
        'ok': status['state'] == 'succeeded' and routed_executor == spec['expected_executor'],
        'task': spec['task'],
        'events': events,
    }


def render_bundle_readme(report):
    lines = [
        '# Mixed Workload Validation Bundle',
        '',
        f"- Run ID: `{report['run_id']}`",
        f"- Generated at: `{report['generated_at']}`",
        f"- Status: `{'succeeded' if report['all_ok'] else 'failed'}`",
        f"- Latest summary: `{report['latest_report_path']}`",
        '',
        '## Scenario drilldowns',
        '',
    ]
    for task in report['tasks']:
        lines.append(f"### {task['name']}")
        lines.append(f"- Expectation: `{task['expectation']['label']}`")
        lines.append(f"- Expected executor: `{task['expectation']['executor']}`")
        lines.append(f"- Routed executor: `{task['observed']['routed_executor']}`")
        lines.append(f"- Final state: `{task['observed']['final_state']}`")
        lines.append(f"- Drilldown JSON: `{task['drilldown']['json_path']}`")
        lines.append('')
    lines.extend(
        [
            '## Shared artifacts',
            '',
            f"- Bundle summary: `{report['bundle']['summary_path']}`",
            f"- Canonical report: `{report['latest_report_path']}`",
        ]
    )
    if report['bundle'].get('audit_log_path'):
        lines.append(f"- Audit log copy: `{report['bundle']['audit_log_path']}`")
    if report['bundle'].get('service_log_path'):
        lines.append(f"- Service log copy: `{report['bundle']['service_log_path']}`")
    lines.append('')
    return '\n'.join(lines)


def materialize_bundle(go_root, bundle_root, run_id, report, state_dir, log_path):
    root = pathlib.Path(go_root)
    bundle_dir = root / bundle_root / run_id
    drilldown_dir = bundle_dir / 'tasks'
    drilldown_dir.mkdir(parents=True, exist_ok=True)

    for task in report['tasks']:
        drilldown_payload = {
            'generated_at': report['generated_at'],
            'run_id': report['run_id'],
            'name': task['name'],
            'task_id': task['task_id'],
            'trace_id': task['trace_id'],
            'expectation': task['expectation'],
            'observed': task['observed'],
            'event_types': task['event_types'],
            'task': task['task'],
            'events': task['events'],
            'ok': task['ok'],
        }
        drilldown_path = drilldown_dir / f"{task['name']}.json"
        write_report(go_root, relpath(drilldown_path, go_root), drilldown_payload)
        task['drilldown'] = {'json_path': relpath(drilldown_path, go_root)}
        task.pop('task', None)
        task.pop('events', None)

    summary_path = bundle_dir / 'summary.json'
    readme_path = bundle_dir / 'README.md'
    report['bundle'] = {
        'path': relpath(bundle_dir, go_root),
        'summary_path': relpath(summary_path, go_root),
        'readme_path': relpath(readme_path, go_root),
    }
    if state_dir:
        copied_audit = copy_artifact(pathlib.Path(state_dir) / 'audit.jsonl', bundle_dir / 'audit.jsonl')
        if copied_audit:
            report['bundle']['audit_log_path'] = relpath(copied_audit, go_root)
    if log_path:
        copied_service_log = copy_artifact(log_path, bundle_dir / 'service.log')
        if copied_service_log:
            report['bundle']['service_log_path'] = relpath(copied_service_log, go_root)

    write_report(go_root, relpath(summary_path, go_root), report)
    readme_path.write_text(render_bundle_readme(report) + '\n', encoding='utf-8')
    return report


def main():
    parser = argparse.ArgumentParser(description='Run a mixed workload matrix against one BigClaw Go control plane')
    parser.add_argument('--go-root', default=str(pathlib.Path(__file__).resolve().parents[2]))
    parser.add_argument('--report-path', default='docs/reports/mixed-workload-matrix-report.json')
    parser.add_argument('--bundle-root', default='docs/reports/mixed-workload-runs')
    parser.add_argument('--timeout-seconds', type=int, default=240)
    parser.add_argument('--autostart', action='store_true', default=True)
    args = parser.parse_args()

    process = None
    log_path = None
    state_dir = None
    report = None
    base_url = os.environ.get('BIGCLAW_ADDR', 'http://127.0.0.1:8080')
    try:
        if args.autostart:
            env, base_url, state_dir = build_autostart_env()
            process, log_path = start_bigclawd(args.go_root, env)
        wait_for_health(base_url)
        timestamp = int(time.time())
        task_specs = default_tasks(timestamp)
        results = []
        for spec in task_specs:
            task = spec['task']
            http_json(base_url + '/tasks', method='POST', payload=task)
            status = wait_task(base_url, task['id'], args.timeout_seconds)
            events = fetch_events(base_url, task['id'])
            results.append(build_task_result(spec, status, events))
        run_id = time.strftime('%Y%m%dT%H%M%SZ', time.gmtime())
        report = {
            'generated_at': time.strftime('%Y-%m-%dT%H:%M:%SZ', time.gmtime()),
            'run_id': run_id,
            'base_url': base_url,
            'state_dir': str(state_dir) if state_dir else '',
            'service_log': str(log_path) if log_path else '',
            'all_ok': all(item['ok'] for item in results),
            'tasks': results,
        }
        report['latest_report_path'] = args.report_path
        report = materialize_bundle(args.go_root, args.bundle_root, run_id, report, state_dir, log_path)
        write_report(args.go_root, args.report_path, report)
        print(json.dumps(report, ensure_ascii=False, indent=2))
        return 0 if report['all_ok'] else 1
    finally:
        if process is not None:
            process.terminate()
            try:
                process.wait(timeout=5)
            except subprocess.TimeoutExpired:
                process.kill()
            if log_path and log_path.exists():
                print(f'bigclawd log: {log_path}', file=sys.stderr)


if __name__ == '__main__':
    raise SystemExit(main())

File: bigclaw-go/scripts/e2e/mixed_workload_matrix.go
package main

import (
	"bytes"
	"encoding/json"
	"flag"
	"fmt"
	"net"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"time"
)

type mixedWorkloadArgs struct {
	GoRoot         string
	ReportPath     string
	TimeoutSeconds int
	Autostart      bool
}

type mixedTaskSpec struct {
	Name             string
	ExpectedExecutor string
	Task             map[string]any
}

type mixedAutostartState struct {
	Env      map[string]string
	BaseURL  string
	StateDir string
}

func main() {
	args := parseMixedWorkloadArgs()
	os.Exit(runMixedWorkloadMatrix(args))
}

func parseMixedWorkloadArgs() mixedWorkloadArgs {
	defaultRoot, err := os.Getwd()
	if err != nil {
		defaultRoot = "."
	}
	args := mixedWorkloadArgs{}
	flag.StringVar(&args.GoRoot, "go-root", defaultRoot, "repo root")
	flag.StringVar(&args.ReportPath, "report-path", "docs/reports/mixed-workload-matrix-report.json", "report output")
	flag.IntVar(&args.TimeoutSeconds, "timeout-seconds", 240, "task timeout in seconds")
	flag.BoolVar(&args.Autostart, "autostart", true, "start a local bigclawd process automatically")
	flag.Parse()
	return args
}

func runMixedWorkloadMatrix(args mixedWorkloadArgs) int {
	goRoot, err := filepath.Abs(args.GoRoot)
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		return 1
	}

	process := (*exec.Cmd)(nil)
	logHandle := (*os.File)(nil)
	logPath := ""
	baseURL := getenvDefault("BIGCLAW_ADDR", "http://127.0.0.1:8080")
	stateDir := ""

	defer func() {
		if process != nil {
			terminateProcess(process, logHandle)
			if logPath != "" {
				fmt.Fprintf(os.Stderr, "bigclawd log: %s\n", logPath)
			}
		}
	}()

	if args.Autostart {
		autostartState, err := buildMixedAutostartEnv()
		if err != nil {
			fmt.Fprintln(os.Stderr, err)
			return 1
		}
		stateDir = autostartState.StateDir
		baseURL = autostartState.BaseURL
		process, logHandle, logPath, err = startMixedBigclawd(goRoot, autostartState.Env)
		if err != nil {
			fmt.Fprintln(os.Stderr, err)
			return 1
		}
	}

	if err := waitForMixedHealth(baseURL, 60, time.Second); err != nil {
		fmt.Fprintln(os.Stderr, err)
		return 1
	}

	timestamp := time.Now().Unix()
	results := make([]map[string]any, 0, 5)
	allOK := true
	for _, spec := range defaultMixedTasks(timestamp) {
		if err := postMixedTask(baseURL, spec.Task); err != nil {
			fmt.Fprintln(os.Stderr, err)
			return 1
		}
		status, err := waitMixedTask(baseURL, asString(spec.Task["id"]), args.TimeoutSeconds)
		if err != nil {
			fmt.Fprintln(os.Stderr, err)
			return 1
		}
		events, err := fetchMixedEvents(baseURL, asString(spec.Task["id"]))
		if err != nil {
			fmt.Fprintln(os.Stderr, err)
			return 1
		}
		routed := firstRoutedEvent(events)
		routedExecutor := ""
		routedReason := ""
		if routed != nil {
			payload := asMap(routed["payload"])
			routedExecutor = asString(payload["executor"])
			routedReason = asString(payload["reason"])
		}
		statusMap := asMap(status)
		ok := asString(statusMap["state"]) == "succeeded" && routedExecutor == spec.ExpectedExecutor
		if !ok {
			allOK = false
		}
		results = append(results, map[string]any{
			"name":              spec.Name,
			"task_id":           asString(spec.Task["id"]),
			"trace_id":          asString(spec.Task["trace_id"]),
			"expected_executor": spec.ExpectedExecutor,
			"routed_executor":   routedExecutor,
			"routed_reason":     routedReason,
			"final_state":       asString(statusMap["state"]),
			"latest_event_type": latestEventType(statusMap),
			"events":            events,
			"ok":                ok,
		})
	}

	report := map[string]any{
		"generated_at": time.Now().UTC().Format(time.RFC3339),
		"base_url":     baseURL,
		"state_dir":    stateDir,
		"service_log":  logPath,
		"all_ok":       allOK,
		"tasks":        results,
	}
	if err := writeMixedReport(goRoot, args.ReportPath, report); err != nil {
		fmt.Fprintln(os.Stderr, err)
		return 1
	}
	encoded, err := json.MarshalIndent(report, "", "  ")
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		return 1
	}
	fmt.Println(string(encoded))
	if allOK {
		return 0
	}
	return 1
}

func getenvDefault(key, fallback string) string {
	if value := os.Getenv(key); value != "" {
		return value
	}
	return fallback
}

func buildMixedAutostartEnv() (mixedAutostartState, error) {
	stateDir, err := os.MkdirTemp("", "bigclawd-mixed-state-")
	if err != nil {
		return mixedAutostartState{}, err
	}
	baseURL, httpAddr, err := reserveMixedBaseURL()
	if err != nil {
		return mixedAutostartState{}, err
	}
	env := copyMixedEnv()
	env["BIGCLAW_HTTP_ADDR"] = httpAddr
	env["BIGCLAW_QUEUE_BACKEND"] = getenvMapDefault(env, "BIGCLAW_QUEUE_BACKEND", "sqlite")
	env["BIGCLAW_QUEUE_SQLITE_PATH"] = filepath.Join(stateDir, "queue.db")
	env["BIGCLAW_AUDIT_LOG_PATH"] = filepath.Join(stateDir, "audit.jsonl")
	env["BIGCLAW_SERVICE_NAME"] = getenvMapDefault(env, "BIGCLAW_SERVICE_NAME", "bigclawd-mixed")
	return mixedAutostartState{Env: env, BaseURL: baseURL, StateDir: stateDir}, nil
}

func getenvMapDefault(env map[string]string, key, fallback string) string {
	if value := env[key]; value != "" {
		return value
	}
	return fallback
}

func copyMixedEnv() map[string]string {
	env := map[string]string{}
	for _, entry := range os.Environ() {
		parts := bytes.SplitN([]byte(entry), []byte("="), 2)
		if len(parts) == 2 {
			env[string(parts[0])] = string(parts[1])
		}
	}
	return env
}

func reserveMixedBaseURL() (string, string, error) {
	listener, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		return "", "", err
	}
	addr := listener.Addr().String()
	if err := listener.Close(); err != nil {
		return "", "", err
	}
	return "http://" + addr, addr, nil
}

func startMixedBigclawd(goRoot string, env map[string]string) (*exec.Cmd, *os.File, string, error) {
	logPath, err := os.CreateTemp("", "bigclawd-mixed-*.log")
	if err != nil {
		return nil, nil, "", err
	}
	cmd := exec.Command("go", "run", "./cmd/bigclawd")
	cmd.Dir = goRoot
	cmd.Stdout = logPath
	cmd.Stderr = logPath
	cmd.Env = mixedEnvSlice(env)
	if err := cmd.Start(); err != nil {
		_ = logPath.Close()
		return nil, nil, "", err
	}
	return cmd, logPath, logPath.Name(), nil
}

func mixedEnvSlice(env map[string]string) []string {
	values := make([]string, 0, len(env))
	for key, value := range env {
		values = append(values, key+"="+value)
	}
	return values
}

func terminateProcess(process *exec.Cmd, logHandle *os.File) {
	if process == nil || process.Process == nil {
		return
	}
	_ = process.Process.Signal(os.Interrupt)
	done := make(chan struct{})
	go func() {
		_, _ = process.Process.Wait()
		close(done)
	}()
	select {
	case <-done:
	case <-time.After(5 * time.Second):
		_ = process.Process.Kill()
		<-done
	}
	if logHandle != nil {
		_ = logHandle.Close()
	}
}

func waitForMixedHealth(baseURL string, attempts int, interval time.Duration) error {
	var lastErr error
	for i := 0; i < attempts; i++ {
		payload, err := mixedHTTPJSON(baseURL+"/healthz", http.MethodGet, nil, 30*time.Second)
		if err == nil && asBool(payload["ok"]) {
			return nil
		}
		lastErr = err
		time.Sleep(interval)
	}
	return fmt.Errorf("service did not become healthy: %v", lastErr)
}

func mixedHTTPJSON(url string, method string, payload any, timeout time.Duration) (map[string]any, error) {
	var body *bytes.Reader
	if payload == nil {
		body = bytes.NewReader(nil)
	} else {
		encoded, err := json.Marshal(payload)
		if err != nil {
			return nil, err
		}
		body = bytes.NewReader(encoded)
	}
	req, err := http.NewRequest(method, url, body)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")
	client := &http.Client{Timeout: timeout}
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	var decoded map[string]any
	if err := json.NewDecoder(resp.Body).Decode(&decoded); err != nil {
		return nil, err
	}
	if resp.StatusCode >= 400 {
		return nil, fmt.Errorf("http %d: %v", resp.StatusCode, decoded)
	}
	return decoded, nil
}

func postMixedTask(baseURL string, task map[string]any) error {
	_, err := mixedHTTPJSON(baseURL+"/tasks", http.MethodPost, task, 30*time.Second)
	return err
}

func waitMixedTask(baseURL string, taskID string, timeoutSeconds int) (map[string]any, error) {
	deadline := time.Now().Add(time.Duration(timeoutSeconds) * time.Second)
	for time.Now().Before(deadline) {
		status, err := mixedHTTPJSON(baseURL+"/tasks/"+taskID, http.MethodGet, nil, 30*time.Second)
		if err != nil {
			return nil, err
		}
		if isMixedTerminal(asString(status["state"])) {
			return status, nil
		}
		time.Sleep(500 * time.Millisecond)
	}
	return nil, fmt.Errorf("task %s did not reach terminal state before timeout", taskID)
}

func isMixedTerminal(state string) bool {
	switch state {
	case "succeeded", "dead_letter", "cancelled", "failed":
		return true
	default:
		return false
	}
}

func fetchMixedEvents(baseURL string, taskID string) ([]map[string]any, error) {
	payload, err := mixedHTTPJSON(baseURL+"/events?task_id="+taskID+"&limit=100", http.MethodGet, nil, 30*time.Second)
	if err != nil {
		return nil, err
	}
	return asMapSlice(payload["events"]), nil
}

func firstRoutedEvent(events []map[string]any) map[string]any {
	for _, event := range events {
		if asString(event["type"]) == "scheduler.routed" {
			return event
		}
	}
	return nil
}

func latestEventType(status map[string]any) string {
	latest := asMap(status["latest_event"])
	return asString(latest["type"])
}

func defaultMixedTasks(timestamp int64) []mixedTaskSpec {
	return []mixedTaskSpec{
		{
			Name:             "local-default",
			ExpectedExecutor: "local",
			Task: map[string]any{
				"id":         fmt.Sprintf("mixed-local-%d", timestamp),
				"trace_id":   fmt.Sprintf("mixed-local-%d", timestamp),
				"title":      "Mixed workload local default",
				"entrypoint": "echo local default",
				"metadata": map[string]any{
					"scenario": "mixed-workload",
					"profile":  "local-default",
				},
			},
		},
		{
			Name:             "browser-auto",
			ExpectedExecutor: "kubernetes",
			Task: map[string]any{
				"id":              fmt.Sprintf("mixed-browser-%d", timestamp),
				"trace_id":        fmt.Sprintf("mixed-browser-%d", timestamp),
				"title":           "Mixed workload browser auto-route",
				"required_tools":  []string{"browser"},
				"container_image": "alpine:3.20",
				"entrypoint":      "echo browser via kubernetes",
				"metadata": map[string]any{
					"scenario": "mixed-workload",
					"profile":  "browser-auto",
				},
			},
		},
		{
			Name:             "gpu-auto",
			ExpectedExecutor: "ray",
			Task: map[string]any{
				"id":             fmt.Sprintf("mixed-gpu-%d", timestamp),
				"trace_id":       fmt.Sprintf("mixed-gpu-%d", timestamp),
				"title":          "Mixed workload gpu auto-route",
				"required_tools": []string{"gpu"},
				"entrypoint":     "python -c \"print('gpu via ray')\"",
				"metadata": map[string]any{
					"scenario": "mixed-workload",
					"profile":  "gpu-auto",
				},
			},
		},
		{
			Name:             "high-risk-auto",
			ExpectedExecutor: "kubernetes",
			Task: map[string]any{
				"id":              fmt.Sprintf("mixed-risk-%d", timestamp),
				"trace_id":        fmt.Sprintf("mixed-risk-%d", timestamp),
				"title":           "Mixed workload high-risk auto-route",
				"risk_level":      "high",
				"container_image": "alpine:3.20",
				"entrypoint":      "echo high risk via kubernetes",
				"metadata": map[string]any{
					"scenario": "mixed-workload",
					"profile":  "high-risk-auto",
				},
			},
		},
		{
			Name:             "required-ray",
			ExpectedExecutor: "ray",
			Task: map[string]any{
				"id":                fmt.Sprintf("mixed-required-ray-%d", timestamp),
				"trace_id":          fmt.Sprintf("mixed-required-ray-%d", timestamp),
				"title":             "Mixed workload explicit ray",
				"required_executor": "ray",
				"entrypoint":        "python -c \"print('required ray')\"",
				"metadata": map[string]any{
					"scenario": "mixed-workload",
					"profile":  "required-ray",
				},
			},
		},
	}
}

func writeMixedReport(goRoot string, reportPath string, payload map[string]any) error {
	outputPath := filepath.Join(goRoot, filepath.FromSlash(reportPath))
	if err := os.MkdirAll(filepath.Dir(outputPath), 0o755); err != nil {
		return err
	}
	encoded, err := json.MarshalIndent(payload, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(outputPath, append(encoded, '\n'), 0o644)
}

func asMap(value any) map[string]any {
	if value == nil {
		return map[string]any{}
	}
	if typed, ok := value.(map[string]any); ok {
		return typed
	}
	return map[string]any{}
}

func asMapSlice(value any) []map[string]any {
	items, ok := value.([]any)
	if !ok {
		if typed, ok := value.([]map[string]any); ok {
			return typed
		}
		return nil
	}
	result := make([]map[string]any, 0, len(items))
	for _, item := range items {
		result = append(result, asMap(item))
	}
	return result
}

func asString(value any) string {
	if typed, ok := value.(string); ok {
		return typed
	}
	return ""
}

func asBool(value any) bool {
	if typed, ok := value.(bool); ok {
		return typed
	}
	return false
}
